### DISCUSSION TITLE
MDS3 Metadata Validations in Authentication Server

### DISCUSSION THREAD

**Discussion by ThaminduDilshan** [webauthn4j/webauthn4j#636] [Q&A] [answered]
Hi,

We are planning to use webauthn4j library in one of our products (server) to perform attestation and assertion validations for the **FIDO compliance test tool**. So far we have successfully integrated attestation validations with `null cert path trust worthiness validator`. Now we are trying to expand our server support for FIDO2 metadata service; In particular MDS3.

We have come up with the following **implementation plan** and have several questions regarding the implementation. I’ve been following the `fido-server-conformance-test-app` implementation provided in the `webauthn4j-spring-security` project.

As I understood, the fido compliance test tool expects metadata service validations only during the device/ security key registration and hence we can implement these validations during the webauthn manager validation. We can use the [DefaultCertPathTrustworthinessValidator](https://github.com/webauthn4j/webauthn4j/blob/0.19.0.RELEASE/webauthn4j-core/src/main/java/com/webauthn4j/validator/attestation/trustworthiness/certpath/DefaultCertPathTrustworthinessValidator.java#L36) introduced during the webauthn4j `0.19.0 release`. It can be initialised using the `MetadataBLOBBasedTrustAnchorRepository` which then can be initialised using a list of `MetadataBLOBProvider`. Here onwards we have two options to follow.

1. Use `FidoMDS3MetadataBLOBProvider` with the obtained metadata service endpoints and the downloaded metadata root certificate.
2. Use `LocalFileMetadataBLOBProvider` with the downloaded metadata files from the obtained metadata service endpoints.

I have tested both options and observed the following.

- `FidoMDS3MetadataBLOBProvider` with the url option seems to be operating **very slowly**. This could be affected by network fetches and slowness as well.
- `LocalFileMetadataBLOBProvider` with the downloaded file content seems to be operating fast, but **couldn’t find a way to provide the metadata service root certificate**. Also metadata test cases regarding the mds certificate had failed (F2, F3, F4 & F5).

#### Questions
1. Apart from the above mentioned baseline approach, the server has to fetch these obtained metadata service endpoints once every month (time interval can be decided based on the requirement) and update the cached metadata files. Is this the correct approach?
2. How can I provide the MDS root certification to the `LocalFileMetadataBLOBProvider`?
3. If that’s not possible, is there a faster way to perform MDS3 metadata blob validations with network fetches?

Thanks in advance.

Regards,
Thamindu

**Comment by ynojima** [comment]
> FidoMDS3MetadataBLOBProvider with the url option seems to be operating very slowly. This could be affected by network fetches and slowness as well.

Yes, but it caches the result. Second call will return the result fast. Please invoke `FidoMDS3MetadataBLOBProvider#refresh` during your server start-up.

> Apart from the above mentioned baseline approach, the server has to fetch these obtained metadata service endpoints once every month (time interval can be decided based on the requirement) and update the cached metadata files. Is this the correct approach?

Yes. Please invoke `FidoMDS3MetadataBLOBProvider#refresh` from your scheduler periodically or `FidoMDS3MetadataBLOBProvider` will auto refresh when `nextUpdate` come (but the call wil take time).
Please check following for auto-refresh logic:
https://github.com/webauthn4j/webauthn4j/blob/45826e9106e3b8057e7aa7786d1f8f3c77305ade/webauthn4j-metadata/src/main/java/com/webauthn4j/metadata/CachingMetadataBLOBProvider.java#L30-L51

> How can I provide the MDS root certification to the `LocalFileMetadataBLOBProvider`?

`LocalFileMetadataBLOBProvider` doesn't validate MetadataBLOB signature as it trusts local metadata BLOB file. So it doen't need MDS3 root certificate.

> Also metadata test cases regarding the mds certificate had failed (F2, F3, F4 & F5).

I suppose you are using only one of `LocalFileMetadataBLOBProvider` or `FidoMDS3MetadataBLOBProvider`.
To pass fido conformance test, you need to load metadata statements files downloadable from FIDO conformance test tools, and also need to load metadata statements file from MDS3 test server.
To load from two sources, you need to use `AggregatingTrustAnchorRepository`, which takes more than one `TrustAnchorRepository`.
Please check how it is used:
https://github.com/webauthn4j/webauthn4j-spring-security/blob/master/samples/fido-server-conformance-test-app/src/main/java/com/webauthn4j/springframework/security/webauthn/sample/app/config/WebSecurityBeanConfig.java#L188

**Comment by ThaminduDilshan** [comment]
Hi @ynojima 

Thanks for the quick reply and explanation.

:arrow_right: 
Regarding answers related to `FidoMDS3MetadataBLOBProvider`, we will experiment on how we can integrate this on server startup and post the status.

:arrow_right:
> I suppose you are using only one of `LocalFileMetadataBLOBProvider` or `FidoMDS3MetadataBLOBProvider`.
> To pass fido conformance test, you need to load metadata statements files downloadable from FIDO conformance test tools, and also need to load metadata statements file from MDS3 test server.

Actually I was using a list of `LocalFileMetadataBLOBProvider` objects corresponding to the endpoints obtained from https://mds3.certinfra.fidoalliance.org/. However, earlier I wasn't using metadata statements files downloadable from FIDO conformance test tool. 

By following the sample provided in the above answer, I was able to implement the logic; however still the metadata test cases related to mds certificate are failing (F2, F3, F4 & F5).
I have published the relevant code snipper for easy reference: https://gist.github.com/ThaminduDilshan/bc064e99b5f4467a6eeb7810fc70e61b.

:arrow_right:
On a side question, can you explain why we need to provide both metadata statements downloadable from FIDO conformance test tool and metadata statements file from MDS3 test server? My concern is whether we have to keep both of these MDS providers in the server implementation on the normal day to day operations.

**Comment by ynojima** [comment]
> Actually I was using a list of LocalFileMetadataBLOBProvider objects corresponding to the endpoints obtained from https://mds3.certinfra.fidoalliance.org/. However, earlier I wasn't using metadata statements files downloadable from FIDO conformance test tool.

Some of metadata BLOB files downloadable from MDS3 test server are invalid and FIDO server implementation is expected to validate and reject them,
but `LocalFileMetadataBLOBProvider` trusts local files and `LocalFileMetadataBLOBProvider` doesn't reject them. Please use `FidoMDS3MetadataBLOBProvider` if you need to fetch metadata BLOB.

> On a side question, can you explain why we need to provide both metadata statements downloadable from FIDO conformance test tool and metadata statements file from MDS3 test server? My concern is whether we have to keep both of these MDS providers in the server implementation on the normal day to day operations.

To trust FIDO alliance certificied security key models, you need to use `FidoMDS3MetadataBLOBProvider` only.
To trust a specific security key model, you need to use `LocalFileMetadataBLOBProvider` or `KeyStoreTrustAnchorRepository` (for example, your organization distributes security keys to members to login internal portal, and the portal only need to trust one specific security key model).
Aggregating two TrustAnchorRepository is normaly not required. Fido conformance test tools request it just for testing.

**Comment by ThaminduDilshan** [comment]
> Some of metadata BLOB files downloadable from MDS3 test server are invalid and FIDO server implementation is expected to validate and reject them,
> but LocalFileMetadataBLOBProvider trusts local files and LocalFileMetadataBLOBProvider doesn't reject them. Please use FidoMDS3MetadataBLOBProvider if you need to fetch metadata BLOB.

So does that mean these certain test cases are expected to fail in `LocalFileMetadataBLOBProvider` while they will pass in `FidoMDS3MetadataBLOBProvider`? 
I guess the reason is when refreshing the `FidoMDS3MetadataBLOBProvider`, it will fetch for the updated BLOBs. Is that correct?

> To trust FIDO alliance certificied security key models, you need to use FidoMDS3MetadataBLOBProvider only.
> To trust a specific security key model, you need to use LocalFileMetadataBLOBProvider or KeyStoreTrustAnchorRepository (for example, your organization distributes security keys to members to login internal portal, and the portal only need to trust one specific security key model).
> Aggregating two TrustAnchorRepository is normaly not required. Fido conformance test tools request it just for testing.

Thanks for the good explanation.

**Comment by ynojima** [comment]
> So does that mean these certain test cases are expected to fail in LocalFileMetadataBLOBProvider while they will pass in FidoMDS3MetadataBLOBProvider?

yes.

> I guess the reason is when refreshing the FidoMDS3MetadataBLOBProvider, it will fetch for the updated BLOBs. Is that correct?

`FidoMDS3MetadataBLOBProvider` fetches and validates BLOB JWS signature, but `LocalFileMetadataBLOBProvider` doesn't. It trusts the local file.

**Comment by ThaminduDilshan** [comment]
@ynojima Thanks for the explanation.

> Yes, but it caches the result. Second call will return the result fast. Please invoke FidoMDS3MetadataBLOBProvider#refresh during your server start-up.

To answer above reply, I've done the `FidoMDS3MetadataBLOBProvider` initialization at the server startup time and able to done the validations faster. However I'm observing few test failures.

`[Server-MDS3 Test server support for Metadata Service 3]`
- F-5 Send a valid ServerAuthenticatorAttestationResponse with FULL "packed" attestation for metadata from MDS3 who's metadata service leaf certificate is revoked, and check that serve returns an error

`[Server-ServerAuthenticatorAttestationResponse-Resp-9 Test server processing "tpm" attestation]`
- P-1 Send a valid ServerAuthenticatorAttestationResponse with "tpm" attestation for SHA-256, and check that server succeeds.
- P-3 Send a valid ServerAuthenticatorAttestationResponse with "tpm" attestation pubArea.nameAlg is not matching algorithm used for generate attested.name, and check that server succeeds.

Latter two tests started to fail when the metadata validations were implemented. Stack for the test failure is below.
```
"com.webauthn4j.validator.exception.CertificateException: invalid cert path"
 detailMessage = "invalid cert path"
 cause = {CertPathValidatorException@22816} "java.security.cert.CertPathValidatorException: Path does not chain with any of the trust anchors"
  index = -1
  certPath = null
  reason = {PKIXReason@22824} "NO_TRUST_ANCHOR"
  detailMessage = "Path does not chain with any of the trust anchors"
  cause = null
```

Do you have any idea about these test failures?

**Comment by ynojima** [comment]
> F-5 Send a valid ServerAuthenticatorAttestationResponse with FULL "packed" attestation for metadata from MDS3 who's metadata service leaf certificate is revoked, and check that serve returns an error

Please call `DefaultCertPathTrustworthinessValidator#setFullChainProhibited(true)`

> P-1 Send a valid ServerAuthenticatorAttestationResponse with "tpm" attestation for SHA-256, and check that server succeeds.
> P-3 Send a valid ServerAuthenticatorAttestationResponse with "tpm" attestation pubArea.nameAlg is not matching algorithm used for generate attested.name, and check that server succeeds.

That is a FIDO conformance test tools end's known issue: https://github.com/fido-alliance/conformance-test-tools-resources/issues/618

**Comment by ThaminduDilshan** [comment]
```
Please call DefaultCertPathTrustworthinessValidator#setFullChainProhibited(true)
```
This doesn't resolve the problem. I was already calling it. Also none of the metadata tests except for the first one will pass without the refresh function call.
```
FidoMDS3MetadataBLOBProvider#refresh()
```

**Comment by ThaminduDilshan** [comment, marked as answer]
Hi @ynojima,

```
[Server-MDS3 Test server support for Metadata Service 3]

F-5 Send a valid ServerAuthenticatorAttestationResponse with FULL "packed" attestation for metadata from MDS3 who's metadata service leaf certificate is revoked, and check that serve returns an error
```

Any resolution around this failure? We assume it's something to do with conformance test cases as it involves metadata. But is it possible this is due to library implementation as the validator success even though it should be failing?

**Reply by ynojima** [reply]
I couldn't reproduce the error:
![image](https://user-images.githubusercontent.com/2139672/156189370-d922ad81-9915-4ab4-ab2d-6c9e24ee04c4.png)

Could you compare my samle (https://github.com/webauthn4j/webauthn4j-spring-security/blob/master/samples/fido-server-conformance-test-app/) and your implementation to find the difference?

**Reply by ThaminduDilshan** [reply]
I compared my implementation with the sample and it's almost the same. Only difference is that I use the library provided default `ObjectConverter` while the sample constructs it by providing the object mappers. Not sure whether that's having an impact on the validations.

```java
objectConverter = new ObjectConverter();
```

Below is the payload.
```js
let testSample = {
    "rawId": "mffkidU4mCdCRTeeqEha5bFWCtaAYpyIa4K7F4k9L38",
    "id": "mffkidU4mCdCRTeeqEha5bFWCtaAYpyIa4K7F4k9L38",
    "response": {
        "attestationObject": "o2NmbXRmcGFja2VkZ2F0dFN0bXSjY2FsZyZjc2lnWEYwRAIgUHPGLDkoAnapa01hVXpfES8H5XalC5kUsojIFvMCBlYCIFtj0xKzPk2yUvp4r-uRagvi6narF1tbKcybE4soY4D1Y3g1Y4FZBEUwggRBMIICKaADAgECAgEBMA0GCSqGSIb3DQEBCwUAMIGhMRgwFgYDVQQDDA9GSURPMiBURVNUIFJPT1QxMTAvBgkqhkiG9w0BCQEWImNvbmZvcm1hbmNlLXRvb2xzQGZpZG9hbGxpYW5jZS5vcmcxFjAUBgNVBAoMDUZJRE8gQWxsaWFuY2UxDDAKBgNVBAsMA0NXRzELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAk1ZMRIwEAYDVQQHDAlXYWtlZmllbGQwHhcNMTgwNTIzMTQzOTQzWhcNMjgwNTIwMTQzOTQzWjCBwjEjMCEGA1UEAwwaRklETzIgQkFUQ0ggS0VZIHByaW1lMjU2djExMTAvBgkqhkiG9w0BCQEWImNvbmZvcm1hbmNlLXRvb2xzQGZpZG9hbGxpYW5jZS5vcmcxFjAUBgNVBAoMDUZJRE8gQWxsaWFuY2UxIjAgBgNVBAsMGUF1dGhlbnRpY2F0b3IgQXR0ZXN0YXRpb24xCzAJBgNVBAYTAlVTMQswCQYDVQQIDAJNWTESMBAGA1UEBwwJV2FrZWZpZWxkMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETzpeXqtsH7yul_bfZEmWdix773IAQCp2xvIw9lVvF6qZm1l_xL9Qiq-OnvDNAT9aub0nkUvwgEN4y8yxG4m1RqMsMCowCQYDVR0TBAIwADAdBgNVHQ4EFgQUVk33wPjGVbahH2xNGfO_QeL9AXkwDQYJKoZIhvcNAQELBQADggIBAIfdoJ_4IIkF0S1Bzgmk6dR6XAYbDsPGcEyiQccGCvMnEOf0EVwXDEYvGsVXsR9h6FA04P7vg5Lx9lGBmI1_0QMYBiIeHT4Kyl8FZ3bTMIiOUJ0MFzKHCrc8snrkkL-iDcJP0AriS-SzgMj7TVFjE2_1LwnHWFo7WWBTnmEEivU_-nbVkqelwISE-MH9wgWscmovmIkZ9534teeL1K6rbg4eenjgyu_iHs4PZ6W7nJZ918Vv5EYbZNhREUgZgaKOyKLT3fDRkwE58FL7der8Osd5ltmus2RjjnmAkJnl5Xzc2u30n39QXRVkeX-HCdIBQL9ve03-XRmUL2Q9w3MkPTiXid0UEPYp19DYcZNfunJtYtnvIfYEze6LY6mJpxo7N3s4T3WsdgHa5nJDuN2DbnIX0zxAj00cz-K0KN0G8Bi3hAJPx1fqCZmIgZHYX9hdkCzJu0nXqmdSY4NVtbzSU9vPL49RBhfv2il4P27owGivOv2DTwSWlvUXcOBJ3xVIuWxHZA-WUqXgBwkMwg59kc5AY7Nq0xXuKkRVFrQvkWeMBakce9I1yyMPgK6XnraY7cyUjakLKj5RL6cjMbldmY567gNv8rD90Q86jbO0fCVTSoontEQGxu3reN1C2XAu6IsfCSmLCesA5l_Bssu71jPi0vV4mVB9-7BL8CiWzPscaGF1dGhEYXRhWKVJlg3liA6MaHQ0Fw9kdmBbj-SuuaKGMseZXPO6gx2XY8EAAAAcF3hTQh8oSYKnP5T1nQlCLAAgmffkidU4mCdCRTeeqEha5bFWCtaAYpyIa4K7F4k9L3-lAQIDJiABIVggKjoFchUbDMkEdDwRVLd13NLqXjV_5wLmu0_-UI-jdjAiWCA4Pl5564DDXGhFn6sriG1ibZmUyRDKPkZbFaNM0aw_UqA",
        "clientDataJSON": "eyJvcmlnaW4iOiJodHRwczovL2xvY2FsaG9zdDo0MDAwIiwiY2hhbGxlbmdlIjoia2lxWlhvTTIwLU5DOXl2UjV2d2JuSzMxUW5wakVjNXpSUmYxY2hNWFFrSSIsInR5cGUiOiJ3ZWJhdXRobi5jcmVhdGUifQ"
    },
    "type": "public-key"
}
```

However I'm running the test tool version `v1.6.42`

---

Even below `ObjectConverter` implementation didn't resolve the issue. F5 test is still failing.

```java
ObjectMapper jsonMapper = new ObjectMapper();
jsonMapper.registerModule(new WebAuthnMetadataJSONModule());
jsonMapper.registerSubtypes(new NamedType(RegistrationExtensionClientInput.class));
ObjectMapper cborMapper = new ObjectMapper(new CBORFactory());
cborMapper.registerSubtypes(new NamedType(AuthenticationExtensionAuthenticatorOutput.class));
objectConverter = new ObjectConverter(jsonMapper, cborMapper);
```

Tests are still failing with `v1.6.37` as well

**Reply by ynojima** [reply]
Do you call `fidoMDS3MetadataBLOBProvider.setRevocationCheckEnabled(false);` ?
https://github.com/webauthn4j/webauthn4j-spring-security/blob/master/samples/fido-server-conformance-test-app/src/main/java/com/webauthn4j/springframework/security/webauthn/sample/app/config/WebSecurityBeanConfig.java#L170

**Reply by ThaminduDilshan** [reply]
Yes. 

To provide more context, I've been using `webauthn4j` library and not the `spring security` framework. Is that possible this has handled from the spring framework itself?

**Reply by ynojima** [reply]
Umm, I recalled what is happening around https://github.com/webauthn4j/webauthn4j-spring-security/blob/master/samples/fido-server-conformance-test-app/src/main/java/com/webauthn4j/springframework/security/webauthn/sample/app/config/WebSecurityBeanConfig.java#L170

As the test case expects, MDS3 Blob revocation status must be verified, but the above configuration disables it because when the check is eanbled, we face another issue: https://github.com/fido-alliance/conformance-test-tools-resources/issues/633

So, what we need to do is to wait for https://github.com/fido-alliance/conformance-test-tools-resources/issues/633 resolution and enable revocation check again by calling `fidoMDS3MetadataBLOBProvider.setRevocationCheckEnabled(true);`

**Reply by ThaminduDilshan** [reply]
Okay. I have removed this line and ran the tests again. Now the MDS F5 test is passing and P1 started to fail.
So I believe these 3 tests must be fixed from FIDO end and there's nothing much we can do.
@ynojima thanks for your great support for getting this resolved.
